signwriting_animation/scripts/minimal_loop.py

```diff
@@ -0,0 +1,265 @@
+from signwriting_animation.data.data_loader import DynamicPosePredictionDataset
+from signwriting_animation.diffusion.core.models import SignWritingToPoseDiffusion
+
+
+def _to_dense(x):
+    """
+    Safely convert a batch tensor to dense float32:
+    - If it's a pose-format MaskedTensor, call .zero_filled() to obtain a dense tensor.
+    - If it's a sparse tensor, densify it.
+    - Cast to float32 and make memory contiguous.
+    """
+    if hasattr(x, "zero_filled"):
+        x = x.zero_filled()
+    if getattr(x, "is_sparse", False):
+        x = x.to_dense()
+    if x.dtype != torch.float32:
+        x = x.float()
+    return x.contiguous()
+
+def sanitize_btjc(x):
+    """Ensure tensor is [B,T,J,C]. Handle sparse or [B,T,P,J,C] inputs."""
+    x = _to_dense(x)
+    if x.dim() == 5:  # [B,T,P,J,C]
+        x = x[:, :, 0, ...]
+    if x.dim() != 4:
+        raise ValueError(f"Expected 4D tensor [B,T,J,C], got {tuple(x.shape)}")
+    return x
+
+
+def btjc_to_bjct(x):
+    """Convert [B, T, J, C] → [B, J, C, T] (model forward format)."""
+    return x.permute(0, 2, 3, 1).contiguous()
+
+
+def bjct_to_btjc(x):
+    """Convert [B, J, C, T] → [B, T, J, C] (loss/metrics format)."""
+    return x.permute(0, 3, 1, 2).contiguous()
+
+
+def masked_mse(pred, tgt, mask_bt):
+    pred, tgt = sanitize_btjc(pred), sanitize_btjc(tgt)
+
+    Tm = min(pred.size(1), tgt.size(1), mask_bt.size(1))
+    pred = pred[:, :Tm]
+    tgt  = tgt[:,  :Tm]
+
+    m4 = mask_bt[:, :Tm].float()[:, :, None, None]   # [B,T,1,1]
+
+    diff2 = (pred - tgt) ** 2                        # [B,T,J,C]
+    num = (diff2 * m4).sum()
+    den = (m4.sum() * pred.size(2) * pred.size(3)).clamp_min(1.0)
+    return num / den
+
+def masked_dtw(pred_btjc, tgt_btjc, mask_bt):
+    pred = sanitize_btjc(pred_btjc)  # [B,T,J,C]
+    tgt  = sanitize_btjc(tgt_btjc)
+    B, T, J, C = pred.shape
+    vals = []
+
+    for b in range(B):
+        t = min(int(mask_bt[b].sum().item()), pred.size(1), tgt.size(1))
+        if t <= 1:
+            continue
+
+        x = pred[b, :t].reshape(t, -1).to(dtype=torch.float32)  # [t, J*C]
+        y = tgt[b,  :t].reshape(t, -1).to(dtype=torch.float32)
+
+        device = x.device
+        D = torch.cdist(x, y)                      # [t, t]
+        Cmat = torch.empty((t, t), device=device)
+        Cmat[0, 0] = D[0, 0]
+        for i in range(1, t):
+            Cmat[i, 0] = D[i, 0] + Cmat[i-1, 0]
+        for j in range(1, t):
+            Cmat[0, j] = D[0, j] + Cmat[0, j-1]
+
+        for i in range(1, t):
+            for j in range(1, t):
+                m = torch.minimum(
+                        torch.minimum(Cmat[i-1, j], Cmat[i, j-1]),
+                        Cmat[i-1, j-1]
+                    )
+                Cmat[i, j] = D[i, j] + m
+
+        vals.append(Cmat[t-1, t-1] / (2 * t))
+
+    if not vals:
+        return torch.tensor(0.0, device=pred.device)
+    return torch.stack(vals).mean()
```

Comment:
we have metrics for poses in this repo: https://github.com/sign-language-processing/pose-evaluation/
including some examples in [here](https://github.com/sign-language-processing/pose-evaluation/blob/main/pose_evaluation/examples/example_metric_construction.py) - feel free to open an issue if you are struggling